File: DynamicProgramming/SWEA1952.py
T = int(input())
for test_case in range(1,1+T):
    prices = list(map(int,input().split()))
    # 0 1일권, 1 1개월권, 2 3개월권, 3 1년권
    days = [0] + list(map(int,input().split()))
    # days[1] = 1월달 이용할 날
    DP = [[3000,3000,3000] for _ in range(13)]
    DP[12] = [days[12]*prices[0], prices[1], 3000]
    DP[11][0] = min(DP[12][0],DP[12][1]) + days[11]*prices[0]
    DP[11][1] = min(DP[12][0],DP[12][1]) + prices[1]
    DP[11][2] = 3000
    DP[10][0] = min(DP[11][0],DP[11][1]) + days[10]*prices[0]
    DP[10][1] = min(DP[11][0],DP[11][1]) + prices[1]
    DP[10][2] = prices[2]
    for month in range(9,0,-1):
        DP[month][0] = min(DP[month+1]) + days[month]*prices[0]
        DP[month][1] = min(DP[month+1]) + prices[1]
        DP[month][2] = min(DP[month+3]) + prices[2]
    answer = min(DP[1]+[prices[3]])
    print(f"#{test_case} {answer}")

File: Backtracking/SWEA1952.java
package Backtracking;

import java.util.Scanner;

public class SWEA1952 {
	static int[] prices = new int[4];
	static int[] days = new int[14];
	static int answer;
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int T = sc.nextInt();
		for (int tc = 1; tc <= T; tc++) {
			for (int i = 0; i < 4; i++) {
				prices[i] = sc.nextInt();
			}
			for (int i = 1; i < 13; i++) {
				days[i] = sc.nextInt();
			}
			answer = prices[3];
			Backtracking(1,0);
			System.out.printf("#%d %d\n",tc,answer);
		}
	}
	
	public static void Backtracking(int month, int totalPrice) {
		if (answer < totalPrice) {
			return;
		}
		if (month == 13) {
			if (answer > totalPrice) {
				answer =totalPrice;
			}
			return;
		}
		if (prices[0]*days[month] < prices[1]) {
			Backtracking(month+1,totalPrice+prices[0]*days[month]);
		}
		else {
			Backtracking(month+1,totalPrice+prices[1]);
		}
		
		if (month <= 10) {
			Backtracking(month+3,totalPrice+prices[2]);
		}
	}
}
